File: front-end/src/app/domain/models/game/action/impl/money/RentCollectionAction.ts
import {Player} from "../../../Player";
import {OwnableLocation} from "../../../board/location/OwnableLocation";
import {MoneyExchangeAction} from "./MoneyExchangeAction";
import {LocationUtil} from "../../../board/location/LocationUtil";

/**
 * A representation, of the action, of rent collection by an owner of an OwnableLocation within a GameSession,
 * from a Player that lands on it.
 *
 * @author Hamza el Haouti
 */
export class RentCollectionAction extends MoneyExchangeAction {
  public constructor(
    id: number,
    renter: Player,
    performedAt: Date,
    amount: number,
    property: OwnableLocation,
    landLord: Player
  ) {
    super(RentCollectionAction.name, id, renter, performedAt, landLord, amount);
    this._location = property;
  }

  private _location: OwnableLocation;

  public get location(): OwnableLocation {
    return this._location;
  }

  public static trueCopy(action: RentCollectionAction): RentCollectionAction | null {
    return new RentCollectionAction(
      action.id,
      Player.trueCopy(action.actor),
      new Date(action.performedAt),
      action.amount,
      <OwnableLocation>LocationUtil.trueCopy(action.location),
      Player.trueCopy(action.recipient),
    );
  }

  public override toJson(): object {
    return Object.assign(
      super.toJson(), {
        "location": this.location.toJson(),
      }
    );
  }

  public override toString(): string {
    return `${this.recipient.user.username} collected $${this.amount} worth of rent, from ${this.actor.user.username}`
      + `, for landing on ${this.location.toString()}`;
  }

}

File: back-end/src/main/java/nl/hva/backend/domain/models/game/action/impl/money/moneyExchangeImpl/RentCollectionAction.java
package nl.hva.backend.domain.models.game.action.impl.money.moneyExchangeImpl;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import nl.hva.backend.domain.models.game.Player;
import nl.hva.backend.domain.models.game.action.impl.money.MoneyExchangeAction;
import nl.hva.backend.domain.models.game.board.location.OwnableLocation;

import javax.persistence.Entity;
import javax.persistence.ManyToOne;

/**
 * A representation, of the action, of rent collection by an owner of an OwnableLocation within a GameSession,
 * from a Player that lands on it.
 *
 * @author Hamza el Haouti
 */
@Entity
@Getter
@Setter
@NoArgsConstructor
public class RentCollectionAction extends MoneyExchangeAction {

    @ManyToOne
    private OwnableLocation location;

    /**
     * Constructor for the manual creation of an instance, that is intended to be persisted to JPA.
     *
     * @param renter   A Player instance, that landed on owned location.
     * @param landLord A player instance, that owns that the location.
     * @param amount   The rent, to be subtracted from the renter, and added to the balance of the landLord.
     */
    public RentCollectionAction(
            Player renter,
            Player landLord,
            double amount,
            OwnableLocation location
    ) {
        super(renter, landLord, amount);
        this.location = location;
    }

}